lib/ReadsUtils/ReadsUtilsImpl.py

```diff
@@ -642,6 +668,223 @@ def get_fq_stats(self, reads_object, file_path):
+        """
+        _download_web_file: download reads source file from web
+
+        file_url: file URL
+        download_type: one of ['Direct Download', 'FTP', 'DropBox', 'Google Drive']
+        rev_file: optional, default as False. Set to True if file_url is for rev_file
+
+        return: file path of downloaded web file
+
+        """
+
+        # prepare local copy file path for copy
+        tmp_file_name = 'tmp_rev_fastq.fastq' if rev_file else 'tmp_fwd_fastq.fastq'
+        dstdir = os.path.join(self.scratch, 'tmp')
+        if not os.path.exists(dstdir):
+            os.makedirs(dstdir)
+        copy_file_path = os.path.join(dstdir, tmp_file_name)
+
+        self._download_file(download_type, file_url, copy_file_path)
+
+        return copy_file_path
+
+
+    def _download_file(self, download_type, file_url, copy_file_path):
+        """
+        _download_file: download execution distributor
+
+        params:
+        download_type: download type for web source file
+        file_url: file URL
+        copy_file_path: output file saving path
+        
+        """
+        if download_type == 'Direct Download':
+            self._download_direct_download_link(file_url, copy_file_path)
+        elif download_type == 'DropBox':
+            self._download_dropbox_link(file_url, copy_file_path)
+        elif download_type == 'FTP':
+            self._download_ftp_link(file_url, copy_file_path)
+        elif download_type == 'Google Drive':
+            self._download_google_drive_link(file_url, copy_file_path)
+
+    def _download_direct_download_link(self, file_url, copy_file_path):    
+        """
+        _download_direct_download_link: direct download link handler 
+
+        params:
+        file_url: direct download URL
+        copy_file_path: output file saving path
+
+        """
+
+        try: online_file = urllib2.urlopen(file_url)
+        except urllib2.HTTPError as e:
+            raise ValueError("The server couldn\'t fulfill the request.\n(Is link publicaly accessable?)\nError code: %s" % e.code)
+        except urllib2.URLError as e:
+            raise ValueError("Failed to reach a server\nReason: %s" % e.reason)
+        else:
+            with closing(online_file):
+                with open(copy_file_path, 'wb') as output:
+                    shutil.copyfileobj(online_file, output)
+
+    def _download_dropbox_link(self, file_url, copy_file_path):
+        """
+        _download_dropbox_link: dropbox download link handler
+                                file needs to be shared publicly 
+
+        params:
+        file_url: dropbox download link
+        copy_file_path: output file saving path
+
+        """
+        # translate dropbox URL for direct download
+        if "?" not in file_url:
+            force_download_link = file_url + '?raw=1'
+        else:
+            force_download_link = file_url.partition('?')[0] + '?raw=1'
+
+        try: online_file = urllib2.urlopen(force_download_link)
```

Comment:
This part of the fn is identical to the last part of the direct download fn, could make a subfn